File: leetcode/5. Longest Palindromic Substring/main.go
package main

import "strings"

func longestPalindrome(s string) string {
	T := "^#" + strings.Join(strings.Split(s, ""), "#") + "#$"
	n := len(T)
	P := make([]int, n)
	C, R := 0, 0

	for i := 1; i < n-1; i++ {
		if R > i {
			P[i] = min(R-i, P[2*C-i])
		}
		for T[i+1+P[i]] == T[i-1-P[i]] {
			P[i]++
		}
		if i+P[i] > R {
			C, R = i, i+P[i]
		}
	}

	maxLen := 0
	centerIndex := 0

	for i, v := range P {
		if v > maxLen {
			maxLen = v
			centerIndex = i
		}
	}

	return s[(centerIndex-maxLen)/2 : (centerIndex+maxLen)/2]
}

func min(a, b int) int {
	if a < b {
		return a
	}

	return b
}

File: leetcode/5. Longest Palindromic Substring/main.ts
const longestPalindrome = (s: string): string => {
    let T = "^#" + s.split("").join("#") + "#$";
    let n = T.length;

    let P = new Array(n).fill(0);
    let C = 0, R = 0;

    for (let i = 1; i < n - 1; i++) {
        P[i] = (R > i) ? Math.min(R - i, P[2 * C - i]) : 0;
        while (T[i + 1 + P[i]] == T[i - 1 - P[i]])
            P[i]++;

        if (i + P[i] > R) {
            C = i;
            R = i + P[i];
        }
    }

    let maxLen = Math.max(...P);
    let centerIndex = P.indexOf(maxLen);

    return s.substring((centerIndex - maxLen) / 2, (centerIndex + maxLen) / 2);
};
